### PR TITLE
Simplified CR convertor

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Enhancement / new feature

### Description

This **DRAFT** PR provides slightly simplified convertor for the custom resources to the new `v1beta2` version. Given the time pressure to release Strimzi 0.22 it cuts down some functionality in order to save development, testing and documentation efforts. 

The file conversion should be mostly done. But there are still many TODOs:
* Converting Kubernetes resources
* Handle properly different error situation
* Improve the tests

### Checklist

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Comment by scholzj** [comment]
@tombentley There is still a work to be done on tests. But I think the other parts of the code should be ready. If you would have some time I would appreciate if you can have a quick look to see if you have any major comments.

File: api-conversion/README.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+You can also list the help for the file conversion:
+
+```
+> bin/api-conversion.sh help convert-file
+Usage: bin/api-conversion.sh convert-file [-d] -f=<inputFile> [-ll=<level>]
+       [-o=<outputFile> | [--in-place]]
+Convert Custom Resources from YAML file
+  -d, --debug              Use debug
+  -f, --file=<inputFile>   The YAML file with the Strimzi Custom Resource which
+                             should be converted
```
Comment (PaulRMellor) [inline api-conversion/README.md:61]:
```suggestion
  -f, --file=<inputFile>   The YAML file for the Strimzi custom resource being converted
```

File: api-conversion/README.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      --in-place           Apply the changes directly to the input file
+                             specified by --file
+      -ll, --log-level=<level>
+                           Set log level to enable logging
+  -o, --output=<outputFile>
+                           The output YAML file with the converted Strimzi
+                             Custom Resource
```
Comment (PaulRMellor) [inline api-conversion/README.md:68]: Creates an output YAML file for the converted custom resource.

File: api-conversion/README.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+```
+> bin/api-conversion.sh help
+Usage: bin/api-conversion.sh [-hV] [COMMAND]
+Conversion tool for Strimzi Custom Resources
+  -h, --help      Show this help message and exit.
+  -V, --version   Print version information and exit.
+Commands:
+  help                  Displays help information about the specified command
+  convert-file, cf      Convert Custom Resources from YAML file
```
Comment (PaulRMellor) [inline api-conversion/README.md:21]:
```suggestion
  convert-file, cf      Convert custom resources from a YAML file
```

File: api-conversion/README.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  convert-resource, cr  Convert Custom Resources directly in Kubernetes
```
Comment (PaulRMellor) [inline api-conversion/README.md:22]:
```suggestion
  convert-resource, cr  Convert custom resources directly in Kubernetes
```

File: api-conversion/README.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
+# API Conversion CLI Tool
+
+With Strimzi's CRD evolving, there needs to be a way to easily migrate existing CR configuration to the new CRD schema.
+That is why we introduced the `v1beta2` version of the schema which removes several deprecated fields.
+To make it as easy as possible to convert the existing custom resources, we provide this conversion tool.
```
Comment (PaulRMellor) [inline api-conversion/README.md:5]:
```suggestion
To make it as easy as possible to convert existing Strimzi custom resources, a conversion tool is provided.
```

File: api-conversion/README.md (hunk 2 above)
Comment (PaulRMellor) [inline api-conversion/README.md:68]:
```suggestion
                          Creates an output YAML file for the converted custom resource.
```

File: api-conversion/README.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+## Converting YAML files with custom resources
+
+One way to use the tool is to convert YAML files. 
+The converted YAML files can be applied to your Kubernetes cluster either directly or for example using GitOps mechanism.
+The input YAML file is specified using the `--file` option.
+
+The converted YAML will be by default printed to standard output.
+You can use the additional options `--in-place` to write the converted YAML into the original source file.
```
Comment (PaulRMellor) [inline api-conversion/README.md:32]:
```suggestion
You can use the `--in-place` option to write the converted YAML into the original source file.
```

File: api-conversion/README.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+You can also specify another output file using the `--output` option.
```
Comment (PaulRMellor) [inline api-conversion/README.md:33]:
```suggestion
Alternatively, you can create a new YAML file for the conversion using the `--output` option.
```

File: api-conversion/README.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+If the input YAL also contains any other Kubernetes resources, they will be kept in the output file without being modified.
+
+You can also list the help for the file conversion:
+
+```
+> bin/api-conversion.sh help convert-file
+Usage: bin/api-conversion.sh convert-file [-d] -f=<inputFile> [-ll=<level>]
+       [-o=<outputFile> | [--in-place]]
+Convert Custom Resources from YAML file
+  -d, --debug              Use debug
```
Comment (PaulRMellor) [inline api-conversion/README.md:59]:
```suggestion
  -d, --debug              Runs the tool in debug mode
```

File: api-conversion/README.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+  -f, --file=<inputFile>   The YAML file with the Strimzi Custom Resource which
+                             should be converted
+      --in-place           Apply the changes directly to the input file
+                             specified by --file
+      -ll, --log-level=<level>
+                           Set log level to enable logging
```
Comment (PaulRMellor) [inline api-conversion/README.md:65]:
```suggestion
                           Sets the log level to enable logging
```

File: api-conversion/LICENSE [hunk 10]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (tombentley) [inline api-conversion/LICENSE:1]: Why do we need a licence file here?

File: api-conversion/README.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+# Read resources from input.yaml, convert it, and store it in input.yaml
+bin/api-conversion.sh convert-file --file input.yaml --in-place
+
+# Read resources from input.yaml, convert it, and store it in output.yaml
+bin/api-conversion.sh convert-file --file input.yaml --output output.yaml
+```
+
+The convertor supports multi-document YAMLs.
+When your input YAML contains multiple Strimzi custom resources, they will be all converted.
+If the input YAL also contains any other Kubernetes resources, they will be kept in the output file without being modified.
```
Comment (tombentley) [inline api-conversion/README.md:50]:
```suggestion
If the input YAML also contains any other Kubernetes resources, they will be kept in the output file without being modified.
```

File: api-conversion/README.md (hunk 8 above)
Comment (tombentley) [reply api-conversion/README.md:59]: Is this "debug mode" as in "attach a debugger" or debug logging, or something else?

File: api-conversion/README.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+                           The output YAML file with the converted Strimzi
+                             Custom Resource
+```
+
+## Converting Kubernetes resources
+
+You can also use the tool to convert Strimzi custom resources directly in your Kubernetes cluster.
+
+You can use the `--kind` option to specify one or more kinds of Strimzi resources.
+If you do not specify any Kind, the tool will convert resources of all kinds.
```
Comment (tombentley) [inline api-conversion/README.md:76]:
```suggestion
If you do not specify any Kind, the tool will convert Strimzi resources of all kinds.
```

File: api-conversion/src/main/java/io/strimzi/kafka/api/conversion/cli/ConvertResourceCommand.java [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+    /**
+     * Reads the resources from the Kubernetes API and converts them
+     */
+    @Override
+    public void run() {
+        try {
+            String namespace;
+            boolean allNamespaces;
+            client = new DefaultKubernetesClient();
+            Crds.registerCustomKinds();
```
Comment (tombentley) [inline api-conversion/src/main/java/io/strimzi/kafka/api/conversion/cli/ConvertResourceCommand.java:199]: Probably better to do this in a static initializer, so they're definitely only registered once.

**Review by tombentley — COMMENTED** [review]
LGTM so far @scholzj

File: api-conversion/LICENSE (hunk 10 above)
Comment (scholzj) [reply api-conversion/LICENSE]: That was used to package the LICENSE with the ZIP / TAR archives. I removed it and copy the license instead from the root directory of the project.

File: api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/Conversion.java [hunk 14]
```diff
@@ -252,6 +266,35 @@ public void convert(T from) {
...
+                    Property property = pu.getProperty(targetClass, name);
+
+                    if (property == null || !property.isReadable()) {
+                        return null;
+                    }
+
+                    Object value = property.get(target);
+
+                    if (value == null) {
+                        return null;
```
Comment (tombentley) [inline api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/Conversion.java:285]: Is it necessary to distinguish the cases of a path whose final property is null from the case where the complete path didn't exist?

File: api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/APIConversionFailedException.java [hunk 15]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.api.conversion.converter;
+
+/**
+ * Exception used to indicate an conversion issue
+ */
+public class APIConversionFailedException extends RuntimeException {
```
Comment (tombentley) [inline api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/APIConversionFailedException.java:10]: Personally I prefer the capitalization `ApiConversionFailed`, but I feel I'm fighting a losing battle with the community on this.

**Review by tombentley — APPROVED** [review]
A couple of nits, but LGTM.

File: api-conversion/README.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+       [-o=<outputFile> | [--in-place]]
+Convert Custom Resources from YAML file
+  -d, --debug              Runs the tool in debug mode
+  -f, --file=<inputFile>   Specifies the YAML file for the custom resource being converted
+      --in-place           Applies the changes directly to the input file
+                             specified by --file
+      -ll, --log-level=<level>
+                           Sets the log level to enable logging
+  -o, --output=<outputFile>
+                          Creates an output YAML file for the converted custom resource.
```
Comment (PaulRMellor) [inline api-conversion/README.md:66]:
```suggestion
                          Creates an output YAML file for the converted custom resource
```

File: api-conversion/README.md [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+* Converting Kubernetes resources
+
+You can list the available features using the `help` command:
+
+```
+> bin/api-conversion.sh help
+Usage: bin/api-conversion.sh [-hV] [COMMAND]
+Conversion tool for Strimzi Custom Resources
+  -h, --help      Show this help message and exit.
+  -V, --version   Print version information and exit.
```
Comment (PaulRMellor) [inline api-conversion/README.md:18]:
```suggestion
  -h, --help      Show this help message and exit
  -V, --version   Print version information and exit
```

File: api-conversion/src/main/java/io/strimzi/kafka/api/conversion/cli/ConvertFileCommand.java [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+     * @throws IOException  Throws IOException if conversion to YAML fails
+     */
+    private void handleMultipartResources(YAMLMapper yamlMapper, Writer writer) throws IOException {
+        List<MultipartResource> resources = MultipartConversions.get().getResources();
+
+        for (MultipartResource resource : resources) {
+            writer.write(yamlMapper.writeValueAsString(resource.getResource()));
+        }
+
+        MultipartConversions.remove();
```
Comment (alesj) [inline api-conversion/src/main/java/io/strimzi/kafka/api/conversion/cli/ConvertFileCommand.java:138]: Put this in finally block

File: api-conversion/README.md (hunk 17 above)
Comment (scholzj) [reply api-conversion/README.md:18]: This is generated by the PicoCLI library. I can delete it from the README, but not from the actual output. So I would leave it as it is.

File: api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/MetricsConversion.java [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+            return new MetricsInvertibleFunction<>(convertedType(), holderType);
+        }
+
+        @Override
+        public T apply(T holder) {
+            if (holder == null) {
+                return null;
+            }
+
+            throw new UnsupportedOperationException("Reverse conversion of metrics is not supported");
```
Comment (alesj) [inline api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/MetricsConversion.java:174]: Not going to implement this?

Comment (scholzj) [reply api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/MetricsConversion.java:174]: No, part of the _simplification_. I do not think this is really needed right now.

File: api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/APIConversionFailedException.java (hunk 15 above)
Comment (scholzj) [reply api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/APIConversionFailedException.java]: Renamed.

File: api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/Conversion.java (hunk 14 above)
Comment (scholzj) [reply api-conversion/src/main/java/io/strimzi/kafka/api/conversion/converter/Conversion.java:285]: No, that should not be necessary.

File: api-conversion/src/test/java/io/strimzi/kafka/api/conversion/converter/AbstractConversionTestBase.java [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+package io.strimzi.kafka.api.conversion.converter;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.strimzi.api.annotations.ApiVersion;
+
+import java.util.function.Function;
+
+abstract class AbstractConversionTestBase<V> {
```
Comment (alesj) [inline api-conversion/src/test/java/io/strimzi/kafka/api/conversion/converter/AbstractConversionTestBase.java:14]: Why the json and cr conversion method split?
I guess it doesn't hurt ...

Comment (scholzj) [reply api-conversion/src/test/java/io/strimzi/kafka/api/conversion/converter/AbstractConversionTestBase.java:14]: It just seemed easier that way.It is not like I run into any issues with the original way or something. It makes it easier to handle the different options etc. as well when they are in separate commands.

**Review by alesj — APPROVED** [review]
Looks really good, nice simplifications + lots of tests!

File: api-conversion/README.md [hunk 21]
```diff
@@ -0,0 +1,126 @@
+# API Conversion CLI Tool
+
+With Strimzi's CRD evolving, there needs to be a way to easily migrate existing CR configuration to the new CRD schema.
```
Comment (PaulRMellor) [inline api-conversion/README.md:3]:
```suggestion
With Strimzi's CRD evolving, there needs to be a way to easily migrate existing custom resource configuration to the new CRD schema.
```

File: api-conversion/README.md [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+That is why we introduced the `v1beta2` version of the schema which removes several deprecated fields.
```
Comment (PaulRMellor) [inline api-conversion/README.md:4]:
```suggestion
That is why we introduced the `v1beta2` version of the schema, which removes several deprecated fields.
```

File: api-conversion/README.md [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+  convert-resource, cr  Convert custom resources directly in Kubernetes
+```
+
+## Converting YAML files with custom resources
+
+One way to use the tool is to convert YAML files. 
+The converted YAML files can be applied to your Kubernetes cluster either directly or for example using GitOps mechanism.
+The input YAML file is specified using the `--file` option.
+
+The converted YAML will be by default printed to standard output.
```
Comment (PaulRMellor) [inline api-conversion/README.md:31]:
```suggestion
By default, the converted YAML is printed to standard output.

File: api-conversion/README.md [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+You can use the `--in-place` option to write the converted YAML into the original source file.
+Alternatively, you can create a new YAML file for the conversion using the `--output` option.
+
+Following example commands show how the tool can be used:
```
Comment (PaulRMellor) [inline api-conversion/README.md:35]:
```suggestion
The following example commands show how the tool is used:
```

File: api-conversion/README.md [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+
+# Read resources from input.yaml, convert it, and store it in input.yaml
+bin/api-conversion.sh convert-file --file input.yaml --in-place
+
+# Read resources from input.yaml, convert it, and store it in output.yaml
+bin/api-conversion.sh convert-file --file input.yaml --output output.yaml
+```
+
+The convertor supports multi-document YAMLs.
+When your input YAML contains multiple Strimzi custom resources, they will be all converted.
```
Comment (PaulRMellor) [inline api-conversion/README.md:49]:
```suggestion
When your input YAML contains multiple Strimzi custom resources, they are all converted.
```

File: api-conversion/README.md [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+If the input YAML also contains any other Kubernetes resources, they will be kept in the output file without being modified.
```
Comment (PaulRMellor) [inline api-conversion/README.md:50]:
```suggestion
If the input YAML also contains any other Kubernetes resources, they are kept in the output file without being modified.
```

File: api-conversion/README.md [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+  -o, --output=<outputFile>
+                          Creates an output YAML file for the converted custom resource
+```
+
+## Converting Kubernetes resources
+
+You can also use the tool to convert Strimzi custom resources directly in your Kubernetes cluster.
+
+You can use the `--kind` option to specify one or more kinds of Strimzi resources.
+If you do not specify any Kind, the tool will convert Strimzi resources of all kinds.
```
Comment (PaulRMellor) [inline api-conversion/README.md:74]:
```suggestion
If you do not specify any kind, the tool will convert Strimzi resources of all kinds.
```

File: api-conversion/README.md [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+
+The `--namespace` option can be used to specify a namespace in which the resources should be converted.
```
Comment (PaulRMellor) [inline api-conversion/README.md:76]:
```suggestion
The `--namespace` option targets the custom resources in a specified namespace for conversion.
```

File: api-conversion/README.md [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+You can also use `--all-namespace` to convert resources in the whole Kubernetes cluster.
+When neither `--namespace` nor `--all-namespace` is specified your current namespace will used.
```
Comment (PaulRMellor) [inline api-conversion/README.md:78]:
```suggestion
If neither `--namespace` nor `--all-namespace` is specified your current namespace will used.
```

File: api-conversion/README.md [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+
+If you want to convert only one resource, you can also use the combination of `--name` and `--kind` options.
```
Comment (PaulRMellor) [inline api-conversion/README.md:80]:
```suggestion
If you want to convert only one resource, you can also use a combination of `--name` and `--kind` options.
```

File: api-conversion/README.md [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+
+Following example commands show how the tool can be used:
```
Comment (PaulRMellor) [inline api-conversion/README.md:82]:
```suggestion
The following example commands show how the tool can be used:
```

**Review by PaulRMellor — APPROVED** [review]

**Pull request merged**